### ISSUE TITLE
yum-dump.py is being run by bash when I try to install apache on fedora 22

### ISSUE THREAD

**Issue by tejamupparaju** [chef/chef#3613]
[root@cm-node2 ~]# sudo chef-client
Starting Chef Client, version 12.4.0
resolving cookbooks for run list: ["apache::apache_service"]
Synchronizing Cookbooks:
  - apache
Compiling Cookbooks...
Converging 1 resources
Recipe: apache::apache_service
  * yum_package[httpd] action install

    ================================================================================
    Error executing action `install` on resource 'yum_package[httpd]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '2'
    ---- Begin output of /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
    STDOUT:
    STDERR: /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 33: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 34: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 35: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 36: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 37: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 38: import: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 40: from: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 41: from: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 42: from: command not found
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 46: syntax error near unexpected token `('
    /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 46: `YUM_VER = version.StrictVersion(yum.__version__)'
    ---- End output of /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
    Ran /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 returned 2

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/apache/recipes/apache_service.rb

      2: package "httpd" do
      3:     action:install
      4: end
      5:

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/apache/recipes/apache_service.rb:2:in `from_file'

    yum_package("httpd") do
      action :install
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      package_name "httpd"
      flush_cache {:before=>false, :after=>false}
      declared_type :package
      cookbook_name "apache"
      recipe_name "apache_service"
    end

Running handlers:
[2015-06-29T13:15:42-07:00] ERROR: Running exception handlers
Running handlers complete
[2015-06-29T13:15:42-07:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 9.212153785 seconds
[2015-06-29T13:15:44-07:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2015-06-29T13:15:44-07:00] ERROR: yum_package[httpd] (apache::apache_service line 2) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '2'
---- Begin output of /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
STDOUT:
STDERR: /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 33: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 34: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 35: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 36: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 37: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 38: import: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 40: from: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 41: from: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 42: from: command not found
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 46: syntax error near unexpected token `('
/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py: line 46: `YUM_VER = version.StrictVersion(yum.__version__)'
---- End output of /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
Ran /bin/bash  /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 returned 2
[2015-06-29T13:15:44-07:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)

**Comment by danielsdeleo** [comment]
Since newer versions of RHEL family platforms are shipping with a different default `python` we have to detect which python to use by inspecting the shebang line. What do you have for `which yum`? Presumably, that's a bash script on your system. This is the default `yum` that's shipping with Fedora 22?

**Comment by thommay** [comment]
Oh, i bet `yum` on 22 is a shell script that just calls `dnf`. This could be a real pain.

**Comment by thommay** [comment]
yeah:
```
[fedora@ip-172-31-17-62 ~]$ head /usr//bin/yum
#!/bin/bash
# The yum executable redirecting to dnf from dnf-yum compatible package.
# This is shell script and not python script to avoid python2/3 conflicts.
```

**Comment by stevendanna** [comment]
While we could fix up the detection a bit, I think we need to dig into whether running yum-dump on such a system will even yield the correct results.

**Comment by thommay** [comment]
it doesn't:
```
[fedora@ip-172-31-17-62 ~]$ sudo python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py
Traceback (most recent call last):
  File "/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py", line 36, in <module>
    import yum
ImportError: No module named yum
```
So, we could rearrange the omnibus rpm to depend on yum, or we can just write a dnf provider, which I'm trialling right now

**Comment by tejamupparaju** [comment]
I ran yum-dump.py on fedora 22 and it just ran fine without any error

[root@cm-node1 ~]# cat /etc/redhat-release
Fedora release 22 (Twenty Two)
[root@cm-node1 ~]# python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30
[option installonlypkgs] kernel kernel-devel kernel-source installonlypkg(kernel) installonlypkg(kernel-module) installonlypkg(vm)
0ad 0 0.0.18 1.fc22 x86_64 [] a fedora
0ad-data 0 0.0.18 1.fc22 noarch [] a fedora
0install 0 2.6.1 2.fc21 x86_64 [] a fedora

**Comment by thommay** [comment]
@tejamupparaju had you first installed yum?

**Comment by tejamupparaju** [comment]
@thommay  No i didnt install Yum, I think yum still exists in fed 22, but all the yum calls are redirected to dnf. Look the output below 

yum
Yum command has been deprecated, redirecting to '/usr/bin/dnf '.
See 'man dnf' and 'man yum2dnf' for more information.
To transfer transaction metadata from yum to DNF, run:
'dnf install python-dnf-plugins-extras-migrate && dnf-2 migrate'

You need to give some command
usage: dnf [options] COMMAND

List of Main Commands

autoremove
check-update              Check for available package upgrades
clean                     Remove cached data
distro-sync               Synchronize installed packages to the latest available versions

**Comment by thommay** [comment]
The output I pasted was from a clean FC22 install; the yum python stuff definitely wasn't there until I installed it. Did you upgrade or was this a fresh install?

**Comment by tejamupparaju** [comment]
I just installed fedora 22 workstation and it seems to have the same output

**Comment by dmbeck** [comment]
I can confirm that running:

```shell
python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py
```
On a clean install of Fedora 22 Workstation returns the required results.

**Comment by dac** [comment]
A monkey-patch to fix this issue is to put the following code inside any file in the libraries/ directory:
    class Chef
        class Provider
            class Package 
                class Yum
                    class YumCache
                        def python_bin
                            "/usr/bin/python"
                        end
                    end
                end
            end
        end
    end

this code overwrites dynamically the function that return the python executable, used to run yum-dump.py.

**Comment by RavenB** [comment]
yum on fedora 22 is indeed a file that calls DNF. our workaround was (as we have allot of fedora 22, and yes it is not the best way indeed but until chef knows how to handle dnf) : 
- DNF install yum (which actually will install a yum-deprecated)
- rename yum to yum-dnf (to just keep it somewhere)
- symbolic link yum to yum-deprecated and chef works normal with any yum scripts/installs.
This avoided us to hack all the cookbooks.

**Comment by RavenB** [comment]
if you change yumdump.py to point to yum-deprecated instead of yum then everything works. Yum-deprecated is installed by doing dnf install yum.    
Alternative : install yum-deprecated, rename yum to yum-dnf, symbolic link yum to yum-deprecated (however not recommended and gets rerased every update.).     
Adapting chef to call yum-deprecated on fedora 22 systems would be better.    
Or can we have a dnf provider ?

**Comment by danielsdeleo** [comment]
dnf provider would be the long term goal for sure. It may take some time to implement.

**Comment by RavenB** [comment]
any easy way to point chef to yum-deprecated for fedora 22 (instead of the ln -s trick i mentioned) ?

**Comment by danielsdeleo** [comment]
Patch is in progress here: https://github.com/chef/chef/pull/3774 Not sure if there's an easier way that doesn't involve modifying chef.

**Comment by thommay** [comment]
This ought to be fixed in current chef releases.

**Comment by brianjmurrell** [comment]
What release did this fix first appear in?  I'm running 12.5.1 and still seeing this problem.
